### PR TITLE
[dynamo] Implement `set.__contains__` for `Tensor` as object match of `FakeTensor` 

### LINKED ISSUE #111556: [dynamo] Implement `set.__contains__` for tensors based on object identity
### 🚀 The feature, motivation and pitch

Workaround to https://github.com/pytorch/pytorch/issues/111544 for `set.__contains__` case

### Alternatives

_No response_

### Additional context

Related https://github.com/pytorch/pytorch/issues/111550

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519

### PR DESCRIPTION
**Pull request by jon-chuang**
Fixes https://github.com/pytorch/pytorch/issues/111556

Dynamo implementation of `set.__contains__` previously used `__eq__` match.

But this is wrong when `__eq__` match does not imply `__hash__` match, as is the case for `torch.Tensor`, leading to inconsistent results. See: https://github.com/pytorch/pytorch/issues/111542

Hence implement as Tensor object match i.e. proxy node `'example_value'` FakeTensor match.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Comment by jon-chuang** [comment]
@lezcano, to continue our discussion, basically, as of now, the recommended way to check aliasing for Tensors is:

`vt.as_proxy().node.meta['example_value']`. 

Reasons:
1. `node` is not preserved between in-place ops
2. `source` is not always available for tensors
3. `get_storage` may be aliased (e.g. view) even though Tensor object is not aliased.

We've fixed the previously identified higher-order-op generates new fake tensor for same fx node issue in: https://github.com/pytorch/pytorch/pull/111565.

Hence, I believe this option should be fairly reliable now barring some corner cases, which should be fixed if identified.

**Review by lezcano — APPROVED** [review]
Yep, cool!

File: torch/_dynamo/variables/lists.py [hunk 1]
```diff
@@ -25,29 +26,47 @@
...
+        search = _get_fake_tensor(search)
+
+    found = None
     for x in items:
-        check = BuiltinVariable(operator.eq).call_function(tx, [x, search], {})
-        if result is None:
-            result = check
+        if must_check_tensor_id:
+            if isinstance(x, variables.TensorVariable):
+                if search is _get_fake_tensor(x):  # Object equivalence
```
Comment (lezcano) [inline torch/_dynamo/variables/lists.py:57]: Actually, do we need to set any object equivalence guard here, or do we assume that these were already set somewhere before?

Comment (jon-chuang) [reply torch/_dynamo/variables/lists.py:57]: For tensors, we guard both aliasing and non-aliasing.

**Pull request closed without merging**